### ISSUE TITLE
Undo keystroke calls Redo when no more undos available

### ISSUE THREAD

**Issue by robmcmullen** [robmcmullen/peppy#764]
_[Trac time 20090902 235904Z]_ When all changes have been undone (and the undo toolbar properly shows it being disabled), the undo keystroke causes the Redo action to be performed. Turning on keystroke debugging and pressing the undo character C-/ twice in succession:

```
multikey.py:1361 UserAccelerators.OnKeyDown: char=308, unichar=308, multiplier=1, id=238
multikey.py:1368 UserAccelerators.OnKeyDown: id=976 flag=0 key=CONTROL keycode=308
multikey.py:1572 UserAccelerators.OnMenu: id=323 event=<wx._core.CommandEvent; proxy of <Swig Object of type 'wxCommandEvent *' at 0x7fffeea2d3c0> >
multikey.py:1666 UserAccelerators.foundMenuAction: evt=<class 'wx._core.CommandEvent'> id=323 index=1 repeat=1 FOUND MENU ACTION Redo id=0x2fa1a50 mode=<__builtin__.PythonMode; proxy of <Swig Object of type 'wxStyledTextCtrl *' at 0x2f8c120> > frame=BufferFrame peppy: Window #1 id=0x2ccfed0
multikey.py:1319 UserAccelerators.doMenu: Calling action Redo id=0x2fa1a50 mode=<__builtin__.PythonMode; proxy of <Swig Object of type 'wxStyledTextCtrl *' at 0x2f8c120> > frame=BufferFrame peppy: Window #1 id=0x2ccfed0
multikey.py:1733 UserAccelerators.reset: resetting status text to 
multikey.py:1572 UserAccelerators.OnMenu: id=322 event=<wx._core.CommandEvent; proxy of <Swig Object of type 'wxCommandEvent *' at 0x7fffeea2d3c0> >
multikey.py:1666 UserAccelerators.foundMenuAction: evt=<class 'wx._core.CommandEvent'> id=322 index=1 repeat=1 FOUND MENU ACTION Undo id=0x2fa1a10 mode=<__builtin__.PythonMode; proxy of <Swig Object of type 'wxStyledTextCtrl *' at 0x2f8c120> > frame=BufferFrame peppy: Window #1 id=0x2ccfed0
multikey.py:1319 UserAccelerators.doMenu: Calling action Undo id=0x2fa1a10 mode=<__builtin__.PythonMode; proxy of <Swig Object of type 'wxStyledTextCtrl *' at 0x2f8c120> > frame=BufferFrame peppy: Window #1 id=0x2ccfed0
multikey.py:1733 UserAccelerators.reset: resetting status text to 
```

**Comment by robmcmullen** [comment]
_[Trac time 20091106 175138Z]_ Looks like it's a gtk specific problem, where it seems that the menu system is generating the incorrect menu id when the Undo keystroke is C-something and the Redo keystroke is C-S-something, when the something is a number or symbol.  Doesn't occur when the something is letter.

Windows doesn't exhibit the incorrect behavior.

**Comment by robmcmullen** [comment]
_[Trac time 20091106 195544Z]_ Looks like it is caused by the GTK system not recognizing things like "Ctrl-Shift-/", when instead it recognizes "Ctrl-?".

Is there some way to determine the mapping between shifted symbol and keycode?

**Comment by robmcmullen** [comment]
_[Trac time 20091112 005620Z]_ b7bcc69ae75640677110280a63b3076152fdee7f Refs #764: added --debug-keypress command line option

**Comment by robmcmullen** [comment]
_[Trac time 20091112 005630Z]_ b5723dd4f0d08137c08b42bf1380e7621b7ba4f0 Fixed #764: well, worked around, really.  Doesn't look like it can be fixed on GTK
- changed C-z to be main undo keystroke, C-/ is now the backup keystroke
- C-S-z is now the main redo keystroke, C-S-/ is the backup keystroke
- because C-z and C-S-z are the keystrokes that appear in the menu, it is properly recognized and the C-/ and C-S-/ are also recognized by the keyboard processing
